realm/realm-library/src/main/java/io/realm/RealmObject.java

```diff
@@ -81,7 +82,7 @@
  */
 
 @RealmClass
-public abstract class RealmObject {
+public abstract class RealmObject<E extends RealmObject> {
```

Comment:
Not it is horrible. Fortunately it is only required for RealmObjects as both RealmResults and RealmList has the type information. 

I tried different versions of generics black magic, but AFAIK it isn't possible for an abstract class to return subclass type without it being explicitly defined somewhere (either as a generic like here, or as input parameter to a method)